utils/regression-analysis/src/main.rs

```diff
@@ -0,0 +1,221 @@
+
+        PALLET_NAMES.iter().any(|&name| name == pallet_name)
+    };
+
+    let junit_xml = std::fs::read_to_string(path).unwrap();
+    let test_suites: TestSuites = from_str(&junit_xml).unwrap();
+    let total_time: BTreeMap<_, _> = [(
+        String::from("Total time"),
+        f64::from_str(&test_suites.time).unwrap(),
+    )]
+    .into();
+    let mut result = junit_tree::build_tree(filter, test_suites);
+    result.insert(String::from(TEST_SUITES_TEXT), total_time);
+    result
+}
+
+fn median(values: &[u64]) -> u64 {
+    assert!(!values.is_empty());
+
+    let len = values.len();
+    if len % 2 == 0 {
+        let i = len / 2;
+        values[i - 1] / 2 + values[i] / 2 + values[i - 1] % 2 + values[i] % 2
+    } else {
+        values[len / 2]
+    }
+}
+
+fn collect_data<P: AsRef<Path>>(
+    data_folder_path: P,
+    output_path: P,
+    disable_filter: bool,
+    preallocate: usize,
+) {
+    let mut statistics: BTreeMap<_, BTreeMap<_, Vec<_>>> = BTreeMap::default();
+    for entry in fs::read_dir(data_folder_path).unwrap() {
+        let executions = build_tree(disable_filter, &entry.unwrap().path());
+        for (ref key, ref times) in executions {
+            if !statistics.contains_key(key) {
+                statistics.insert(key.clone(), Default::default());
+            }
+
+            let previous_times = statistics.get_mut(key).unwrap();
+            for (key, &time) in times {
+                let time = (1_000_000_000.0 * time) as u64;
+
+                if let Some(time_vec) = previous_times.get_mut(key) {
+                    let i = match time_vec.binary_search(&time) {
+                        Ok(i) => i,
+                        Err(i) => i,
+                    };
+
+                    time_vec.insert(i, time);
+                } else {
+                    let mut time_vec = Vec::with_capacity(preallocate);
+                    time_vec.push(time);
+
+                    previous_times.insert(key.clone(), time_vec);
+                }
+            }
+        }
+    }
+
+    let writer = std::fs::File::create(output_path).unwrap();
+    serde_json::to_writer_pretty(writer, &statistics).unwrap();
+}
+
+fn compare<P: AsRef<Path>>(data_path: P, current_junit_path: P, disable_filter: bool) {
+    let statistics: BTreeMap<String, BTreeMap<String, Vec<u64>>> =
+        serde_json::from_str(&fs::read_to_string(data_path).unwrap()).unwrap();
+    let executions = build_tree(disable_filter, current_junit_path);
+    let mut compared = executions
+        .iter()
+        .filter_map(|(key, tests)| {
+            statistics.get(key).map(|test_times| {
+                let test_stats = tests
+                    .iter()
+                    .filter_map(|(key, &time)| {
+                        test_times.get(key).map(|times| {
+                            let len = times.len();
+                            let len_remainder = len % 2;
+                            let quartile_lower = median(&times[..len / 2]);
```

Comment:
oh, I didn't find code where `times` sorted could you provide that `.sort()` explicitly here, or that median finding is invalid.

I think it might be better to also add comparison with the average, not only with median

Reply:
Good catch! The array is sorted before serialization but yeah, you're right that after deserialization order may be wrong.

However I disagree to compute an average: the characteristic is not robust regarding to outliers.

Reply:
> However I disagree to compute an average: the characteristic is not robust regarding to outliers.

Yeah, but one more dimension wouldn't be excess, especially for such careful and lightweight task.
Median doesn't show us all situation also, but makes us closer to truth (e.g. [2,3,3,3,6,7] median eqs 3, while average eqs 4, and for this case average seems more realistic)